browser/ui/webui/brave_rewards_source.cc

```diff
@@ -0,0 +1,162 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ui/webui/brave_rewards_source.h"
+
+#include "base/memory/ref_counted_memory.h"
+#include "chrome/browser/bitmap_fetcher/bitmap_fetcher_service.h"
+#include "chrome/browser/bitmap_fetcher/bitmap_fetcher_service_factory.h"
+#include "chrome/browser/profiles/profile.h"
+#include "third_party/skia/include/core/SkBitmap.h"
+#include "ui/gfx/codec/png_codec.h"
+#include "url/gurl.h"
+
+namespace {
+
+typedef base::RepeatingCallback<void(BitmapFetcherService::RequestId request_id,
+                                     const GURL& url,
+                                     const SkBitmap& bitmap)>
+    RewardsResourceFetcherCallback;
+
+// Calls the specified callback when the requested image is downloaded.  This
+// is a separate class instead of being implemented on BraveRewardsSource
+// because BitmapFetcherService currently takes ownership of this object.
+class RewardsResourceFetcherObserver : public BitmapFetcherService::Observer {
+ public:
+  explicit RewardsResourceFetcherObserver(
+      const GURL& url,
+      const RewardsResourceFetcherCallback& rewards_resource_fetcher_callback)
+      : url_(url),
+        rewards_resource_fetcher_callback_(rewards_resource_fetcher_callback) {}
+
+  void OnImageChanged(BitmapFetcherService::RequestId request_id,
+                      const SkBitmap& image) override {
+    DCHECK(!image.empty());
+    rewards_resource_fetcher_callback_.Run(request_id, url_, image);
+  }
+
+ private:
+  GURL url_;
+  const RewardsResourceFetcherCallback rewards_resource_fetcher_callback_;
+
+  DISALLOW_COPY_AND_ASSIGN(RewardsResourceFetcherObserver);
+};
+
+scoped_refptr<base::RefCountedMemory> BitmapToMemory(const SkBitmap* image) {
+  base::RefCountedBytes* image_bytes = new base::RefCountedBytes;
+  gfx::PNGCodec::EncodeBGRASkBitmap(*image, false, &image_bytes->data());
+  return image_bytes;
+}
+
+}  // namespace
+
+BraveRewardsSource::BraveRewardsSource(Profile* profile)
+    : profile_(profile->GetOriginalProfile()) {}
```

Comment:
I know this is really old, but it's still in the code and we shouldn't be doing this. There is a method to "redirect" webuis to open in regular profiles which is used by settings, etc...

Reply:
This isn't a WebUI though, it's a URLDataSource. Our code is based off of upstream's favicon data source implementation, which does a similar thing (though maybe that's also problematic): https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/ui/webui/favicon_source.cc;l=71